content/en/altinity-kb-integrations/ClickHouse_python_drivers.md

```diff
@@ -0,0 +1,152 @@
+
+**Python ecosystem libs/modules:**
+
+- Good Pandas/Numpy support: [https://clickhouse.com/docs/en/integrations/python#consuming-query-results-with-numpy-pandas-or-arrow](https://clickhouse.com/docs/en/integrations/python#consuming-query-results-with-numpy-pandas-or-arrow)
+- Decent SQLAlchemy 1.3 and 1.4 support (limited feature set)
+
+It is the most recent driver with the latest feature set (query context and query streaming …. )
+
+You can check multiple official examples here: 
+
+[https://github.com/ClickHouse/clickhouse-connect/tree/457533df05fa685b2a1424359bea5654240ef971/examples](https://github.com/ClickHouse/clickhouse-connect/tree/457533df05fa685b2a1424359bea5654240ef971/examples)
+
+Also some Altinity examples from repo:
+
+[https://github.com/lesandie/clickhouse-tests/blob/main/scripts/test_ch_connect_asyncio_insert.py](https://github.com/lesandie/clickhouse-tests/blob/main/scripts/test_ch_connect_asyncio_insert.py)
+
+You can clone the repo and use the helper files like `DDL.sql`   to setup some tests.
+
+Clickhouse-connect can use a connection pooler (based on urllib3) [https://clickhouse.com/docs/en/integrations/python#customizing-the-http-connection-pool](https://clickhouse.com/docs/en/integrations/python#customizing-the-http-connection-pool)
+
+### Most common use cases:
+
+#### Managing ClickHouse `session_id`:
+
+- clickhouse-driver
+    - Because it is using the Native Interface `session_id` is managed internally by clickhouse, so it is very rare (unless using asyncio) to get:
+    
+    `Code: 373. DB::Exception: Session is locked by a concurrent client. (SESSION_IS_LOCKED)` . 
+    
+
+- clickhouse-connect: How to use clickhouse-connect in a pythonic way and avoid getting `SESSION_IS_LOCKED`  exceptions:
+    - [https://clickhouse.com/docs/en/integrations/python#managing-clickhouse-session-ids](https://clickhouse.com/docs/en/integrations/python#managing-clickhouse-session-ids)
+    - If you want to specify a session_id per query you should be able to use the setting dictionary to pass a `session_id` for each query (note that ClickHouse will automatically generate a `session_id` if none is provided).
+    
+    ```python
+    SETTINGS = {"session_id": "dagster-batch" + "-" + f"{time.time()}"}
+    client.query("INSERT INTO table ....", settings=SETTINGS)
+    ```
+    
+
+Also in clickhouse documentation some explanatio how to set session_id with another approach: [https://clickhouse.com/docs/en/integrations/python#managing-clickhouse-session-ids](https://clickhouse.com/docs/en/integrations/python#managing-clickhouse-session-ids)
+
+[ClickHouse Connect Driver API | ClickHouse Docs](https://clickhouse.com/docs/en/integrations/language-clients/python/driver-api#common-method-arguments)
```

Comment:
link should be to https://clickhouse.com/docs/en/integrations/python#clickhouse-connect-driver-api, I think.